### PR TITLE
Optimizing slow tests

### PR DESCRIPTION
**Pull request by sleepy-owl**
Related to #3823 

Hi,

I optimized some parameters in several tests. This reduces the running times of these times and would help contribute to reducing the build time. 

Please let me know if you have any other suggestions or any necessary checks that I should do on these tests.

Thanks!

### CODE REVIEW

**Comment by sleepy-owl** [comment]
CC @twiecki

**Comment by ColCarroll** [comment]
Thanks for this contribution -- I think it is super welcome. 

Can you help me understand what is going on with some of the parameters? In particular, I am confused about
- The number of samples seem pretty specific (as in, not powers of 10)
- Sometimes there are multiple chains
- Sometimes there is extra tuning

**Comment by ColCarroll** [comment]
Also, using the statistical test of "looking at it", it looks like it lowers the testing time, or at least doesn't make it slower, but I sort of expected a bigger change (this is a knock on my intuition, not the work here)! 

Does this chart look in line with what you expected, @sleepy-owl? I guess some of the speedups I expected would come from running the tests on a multicore machine with the `chains=n` arguments you added...

image

**Comment by sleepy-owl** [comment]
Hi,

We actually developed a research tool which uses Bayesian optimization to tune hyper-parameters. For that, we need to define the space of hyper-parameters and a scoring function. For parameters like iterations/number of samples/tuning steps, I used a large interval with gaps of 10. For parameters like learning rate, I used an interval from `1e-4` to `0.5`. For the scoring each parameter combination, we log the values in the assertions and use statistical convergence tests like [Geweke Diagnostic](https://pymc-devs.github.io/pymc/modelchecking.html#formal-methods) to determine how many samples to get. Then we compute the probability of the test failure. We accept the parameter combination and use the average runtime of the test as the score if the computed probability is below 0.01. 

We applied this method to a number of tests in pymc3 and were able to reduce the run times for a number of tests. I see that the build time also reduced to 1hr 43 mins (from > 1hr 50mins). If you think this would be useful, we could help reduce more tests in the test suite. :) 

Hope this explanation helps. Regarding chains, we verified that the speedups happen on single cores. What would you recommend for testing? Do you guys usually write the tests using single chains? Also given that the server machine are usually multi-core, does it make sense to actually have multiple chains?

**Comment by michaelosthege** [comment]
I think one key is to discriminate between tests that actually need lots of iterations and tests that could do `tune=10, draws=10` because they just test things like shapes or types.

We do have tests of both kinds.

**Comment by sleepy-owl** [comment]
@ColCarroll wdyt?

**Comment by twiecki** [comment]
I feel like 7 minutes isn't significant enough to warrant the weirder numbers, this would completely change if the speed-up was higher in terms of %. Appreciate the work on this, however!
